lib/HeroBonus.cpp

```diff
@@ -1496,3 +1609,89 @@ void LimiterList::add( TLimiterPtr limiter )
 {
 	limiters.push_back(limiter);
 }
+
+// Updaters
+
+std::shared_ptr<Bonus> Bonus::addUpdater(TUpdaterPtr Updater)
+{
+	updater = Updater;
+	return this->shared_from_this();
+}
+
+IUpdater::~IUpdater()
+{
+}
+
+std::string IUpdater::toString() const
+{
+	return typeid(*this).name();
+}
+
+ScalingUpdater::ScalingUpdater() : valPer20(0), stepSize(1)
+{
+}
+
+ScalingUpdater::ScalingUpdater(int valPer20, int stepSize) : valPer20(valPer20), stepSize(stepSize)
+{
+}
+
+bool ScalingUpdater::update(Bonus & b, const CBonusSystemNode & context) const
+{
+	if(context.getNodeType() == CBonusSystemNode::HERO)
+	{
+		int level = static_cast<const CGHeroInstance &>(context).level;
+		int steps = stepSize ? level / stepSize : level;
+		//rounding follows format for HMM3 creature specialty bonus
+		int newVal = (valPer20 * steps + 19) / 20;
+		if(b.val != newVal)
+		{
+			b.val = newVal;
+			return true;
+		}
+	}
+	return false;
+}
+
+std::string ScalingUpdater::toString() const
+{
+	char buf[100];
+	sprintf(buf, "ScalingUpdater(valPer20=%d, stepSize=%d)", valPer20, stepSize);
+	return std::string(buf);
+}
+
+JsonNode ScalingUpdater::toJsonNode() const
+{
+	JsonNode root(JsonNode::DATA_STRUCT);
+
+	root["type"].String() = "GROWS_WITH_LEVEL";
+	root["parameters"].Vector().push_back(JsonUtils::intNode(valPer20));
+	if(stepSize > 1)
+		root["parameters"].Vector().push_back(JsonUtils::intNode(stepSize));
+
+	return root;
+}
+
+std::string nameForBonus(const Bonus & bonus)

```

Comment:
Why this function is not part of Bonus class?

Reply:
My only excuse is that I must have been sleepier than I thought. Fixed.